mmcv/ops/csrc/onnxruntime/cpu/corner_pool.cpp

```diff
@@ -0,0 +1,157 @@
+#include "corner_pool.h"
+#include "../ort_mmcv_utils.h"
+
+void TopPoolForwardCPU(const float *input, float *output, float *tmp_output,
+                          const int nthreads, const int channels,
+                          const int height, const int width) {
+    int batch_size = nthreads / channels / width / height;
+    for (int n = 0; n < batch_size; n++) {
+        int index_n = n * channels * width * height;
+        for (int c = 0; c < channels; c++) {
+            int index_n_c = index_n + c * width * height;
+            for (int w = 0; w < width; w++) {
+                // copy column from output to tmp_output
+                for (int h = 0; h < height; h++) {
+                    int index = index_n_c + h * width + w;
+                    tmp_output[h] = output[index];
+                }
+                // do top_pool
+                for (int ind = 1; ind < height; ind <<= 1) {
+                    for (int h = 0; h < height - ind; h++) {
+                        output[index_n_c + h * width + w] = std::max(tmp_output[h], tmp_output[h+ind]);
+                    }
+                    // copy column from updated output to tmp_output
+                    for (int h = 0; h < height - ind; h++) {
+                        tmp_output[h] = output[index_n_c + h * width + w];
+                    }
+                } // for ind
```

Comment:
As the corner pool is basically a `Scan` operation on a given dimension. This code seems too complex.
Could you provide a brief description of the logic behind the code?

Reply:
Hi, actually the implementation is modified from [here](https://github.com/open-mmlab/mmcv/blob/master/mmcv/ops/csrc/pytorch/corner_pool.cpp).

Reply:
This looks like some kind of parallel implementation of scan (like Kogge-Stone or Hillis-Steele?). I do not think it is the best choice to implement this ops on the CPU.

Code below can do the very same as `CornerPool('right')` without memcpy, additional temp workspace, and it is much simple.
```python
def cp_right(data): 
     batch,channel,height,width = data.shape 
     output = torch.empty_like(data) 
     for n in range(batch): 
         for c in range(channel): 
            for h in range(height): 
                 output[n,c,h,0] = data[n,c,h,0] 
                 for w in range(1, width): 
                     output[n,c,h,w] = max(output[n,c,h,w], data[n,c,h,w-1]) 
     return output 
```

Reply:
Thanks. Another version with no memcpy will be committed soon.